build_resolvers/test/resolver_test.dart

```diff
@@ -680,4 +681,52 @@ int? get x => 1;
       }, resolvers: AnalyzerResolvers());
     });
   });
+
+  group('astNodeFor', () {
+    test('can return an unresolved ast', () {
+      return resolveSources({
+        'a|web/main.dart': ' main() {}',
+      }, (resolver) async {
+        var lib = await resolver.libraryFor(entryPoint);
+        var unit = await resolver.astNodeFor(lib.topLevelElements.first);
+        expect(unit, isA<AstNode>());
+        expect(unit.toSource(), 'main() {}');
+      }, resolvers: AnalyzerResolvers());
+    });
+
+    test('can return an resolved ast', () {
+      return resolveSources({
+        'a|web/main.dart': 'main() {}',
+      }, (resolver) async {
+        var lib = await resolver.libraryFor(entryPoint);
+        var unit = await resolver.astNodeFor(lib.topLevelElements.first,
+            resolve: true);
+        expect(unit, isA<AstNode>());
+        expect(unit.toSource(), 'main() {}');
+      }, resolvers: AnalyzerResolvers());
+    });
+
+    test(
+        'can get an AstNode for an old Element after resolving additional assets',
+        () async {
+      var resolvers = AnalyzerResolvers();
+      await resolveSources({
+        'a|web/main.dart': 'int x;',
+        'a|web/other.dart': '',
+      }, (resolver) async {
+        var lib = await resolver.libraryFor(entryPoint);
+        var x = lib.topLevelElements.firstWhere((x) => !x.isSynthetic);
+        expect(x.name, 'x');
+        expect(await resolver.isLibrary(AssetId('a', 'web/other.dart')), true);
+
+        // Validate that direct session usage would throw
+        expect(() => lib.session.getParsedLibraryByElement(x.library),
+            throwsA(isA<InconsistentAnalysisException>()));
+
+        var astNode = await resolver.astNodeFor(x);
```

Comment:
We should have a version of this same test that passes `resolve: true` since that is the case we are unsure about.

Reply:
Added, although it doesn't actually check the behavior we care about.

I added an additional test that copies the code and does check that behavior also.